primitive/zk_rollup/src/merkle_circuit/root.rs

```diff
@@ -0,0 +1,191 @@
+    fn default() -> Self {
+        Self {
+            leaf: Default::default(),
+            path: [(Fr::zero(), Fr::zero()); K],
+            root: Default::default(),
+            path_pos: [0; K],
+        }
+    }
+}
+
+impl<const K: usize> MerkleMembershipCircuit<K> {
+    pub(crate) fn new(leaf: Fr, root: Fr, path: [(Fr, Fr); K], path_pos: [u64; K]) -> Self {
+        Self {
+            leaf,
+            path,
+            root,
+            path_pos,
+        }
+    }
+}
+
+fn hash(composer: &mut Builder<TatePairing>, inputs: (Witness, Witness)) -> Witness {
+    let sum = Constraint::default()
+        .left(1)
+        .constant(Fr::ADDITIVE_GENERATOR)
+        .a(inputs.0);
+    let gen_plus_first = composer.gate_add(sum);
+
+    let first_hash = Constraint::default().left(2).a(gen_plus_first);
+    let first_hash = composer.gate_add(first_hash);
+
+    let sum = Constraint::default()
+        .left(1)
+        .constant(Fr::ADDITIVE_GENERATOR)
+        .a(inputs.1);
+
+    let gen_plus_second = composer.gate_add(sum);
+
+    let second_hash = Constraint::default().left(2).a(gen_plus_second);
+    let second_hash = composer.gate_add(second_hash);
+
+    composer.gate_add(
+        Constraint::default()
+            .left(1)
+            .right(1)
+            .a(first_hash)
+            .b(second_hash),
+    )
+}
+
+fn calculate_root<const K: usize>(
+    composer: &mut Builder<TatePairing>,
+    leaf: Fr,
+    path: [(Fr, Fr); K],
+    path_pos: [u64; K],
+) -> Result<Witness, Error> {
+    let mut prev = composer.append_witness(leaf);
+
+    let path: Vec<(Witness, Witness)> = path
+        .iter()
+        .map(|(node_l, node_r)| {
+            (
+                composer.append_witness(*node_l),
+                composer.append_witness(*node_r),
+            )
+        })
+        .collect();
+
+    let path_pos: Vec<Witness> = path_pos
+        .iter()
+        .map(|pos| composer.append_witness(JubjubScalar::from(*pos)))
+        .collect();
+
+    for ((left, right), pos) in path.into_iter().zip(path_pos) {
+        // TODO: If provided leaf == 0, and w1 == 0 || w2 == 0, then we pass both checks with invalid leaf
+        // left ^ prev == 0, if equal
+        let w1 = composer.append_logic_xor(left, prev, 256);
+        // right ^ prev == 0, if equal
+        let w2 = composer.append_logic_xor(right, prev, 256);
+        // if one is 0, then and will result to 0
+        let check = composer.append_logic_and(w1, w2, 256);
+        composer.assert_equal_constant(check, 0, None);
+
+        prev = hash(composer, (left, right));
+    }
+
+    Ok(prev)
+}
+
+pub(crate) fn check_membership<const K: usize>(
+    composer: &mut Builder<TatePairing>,
+    leaf: Fr,
+    root: Fr,
+    path: [(Fr, Fr); K],
+    path_pos: [u64; K],
+) -> Result<(), Error> {
+    let real_root = composer.append_witness(root);
+
+    let root = calculate_root(composer, leaf, path, path_pos)?;
+    composer.assert_equal(root, real_root);
```

Comment:
what's real_root?
new_root or precomputed_root would be better

Reply:
Fixed